tests/jballerina-unit-test/src/test/resources/test-src/types/typereftype/type_reference.bal

```diff
@@ -226,9 +298,34 @@ function getImmutable(ImmutableIntArrayOrStringArray x) returns ImmutableIntArra
     return [];
 }
 
+type AnyDataType anydata;
+
+type JsonType json;
+
+type AnyType any;
+
+function testUnionTypeRefWithMap() {
+    AnyDataType map1 = {"a": 1};
+    JsonType map2 = {"b": 2};
+    AnyType map3 = {"c": 3};
+
+    assertTrue(map1 is anydata);
+    assertEqual(map1.toString(), "{\"a\":1}");
+
+    assertTrue(map2 is json);
+    assertEqual(map2.toString(), "{\"b\":2}");
+
+    assertTrue(map3 is any);
+    assertEqual(map3.toString(), "{\"c\":3}");
+}
+
```

Comment:
The following crashes?
```ballerina
type T6 map<json>;

function init() returns error? {
    T6 t14 = {"a": 149, "b": [{"a": {"b": {"c": 123}}}], "c": null};
    json|error b = (<json[]>(t14?.b))[0]?.a?.b;
}
```

Reply:
Below crashes with the same stacktrace which does not rely on type refernce. This is an unrelated bug it seems. This  fix is done related for type references alone. 
```ballerina
function init() returns error? {
    map<json> t14 = {"a": 149, "b": [{"a": {"b": {"c": 123}}}], "c": null};
    json|error b = (<json[]>(t14?.b))[0]?.a?.b;
}
```

Created https://github.com/ballerina-platform/ballerina-lang/issues/35384